File: leetcode/103/P103.rb
# 103. Binary Tree Zigzag Level Order Traversal
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[][]}
def zigzag_level_order(root)
	valLists = []
	if root == nil
		return valLists
	end
	nodeList = [root]
	valLists << [root.val]
	dirct = false
	while !nodeList.empty?
		valList = []
		newNodeList = []
		nodeList.reverse_each do |node|
			if dirct
				addNode node.left, newNodeList, valList
				addNode node.right, newNodeList, valList
			else
				addNode node.right, newNodeList, valList
				addNode node.left, newNodeList, valList
			end
		end
		dirct = !dirct
		if !valList.empty?
			valLists << valList
		end
		nodeList = newNodeList
	end
	valLists
end

def addNode node, nodeList, valList
	if node
		nodeList << node
		valList << node.val
	end
end

File: leetcode/103/src/P103.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.ListIterator;

/**
 * Created by qwordy on 2/24/16.
 * 103. Binary Tree Zigzag Level Order Traversal
 */

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

class Solution {
	public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
		List<List<Integer>> valLists = new ArrayList<>();
		if (root == null)
			return valLists;
		List<TreeNode> nodeList = Collections.singletonList(root);
		List<Integer> valList = Collections.singletonList(root.val);
		valLists.add(valList);
		boolean dirct = false;
		while (!nodeList.isEmpty()) {
			valList = new ArrayList<>();
			List<TreeNode> newNodeList = new ArrayList<>();
			ListIterator<TreeNode> it = nodeList.listIterator(nodeList.size());
			while (it.hasPrevious()) {
				TreeNode node = it.previous();
				if (dirct) {
					addNode(node.left, newNodeList, valList);
					addNode(node.right, newNodeList, valList);
				} else {
					addNode(node.right, newNodeList, valList);
					addNode(node.left, newNodeList, valList);
				}
			}
			dirct = !dirct;
			if (!valList.isEmpty())
				valLists.add(valList);
			nodeList = newNodeList;
		}
		return valLists;
	}

	private void addNode(TreeNode node, List<TreeNode> nodeList, List<Integer> valList) {
		if (node != null) {
			nodeList.add(node);
			valList.add(node.val);
		}
	}
}

public class P103 {
}
